### PR TITLE
Don't subclass RecData for RecipeManager

### PR DESCRIPTION
**Pull request by takluyver**
RecipeManager being a subclass of RecData was causing problems for the singleton-ish machinery: if the single RecData instance was set up before RecipeManager, code that needed a RecipeManager object could get a RecData object instead. This aims to fix that, by moving from inheritance to composition.

Of course, lots of code uses methods from the (former) parent class on a RecipeManager object. Rather than trying to find and modify every case of that, I've added a `__getattr__` method so it will still work.

I've also understood & cleaned up the logic for the not-quite-singleton machinery in RecData, and replicated it for RecipeManager. The idea is to have one instance of each per SQLite database file (or per SQLalchemy database URL).

### CODE REVIEW

**Review by cydanil — APPROVED** [review]
These are fairly in-depth changes, but that part of the application is heavily used, and from using the changes and perusing the code, it looks good!
Additional thumbs-up for the clear comments in `RecipeManager.__getattr__`.

The only minor comment I have are to break the long lines no. 1965 and 1967  in `gourmet/backends/db.py`.

File: gourmet/backends/db.py [hunk 1]
```diff
@@ -114,7 +114,15 @@ class DBObject:
 # categories_table: id -> recipe_id, category_entry_id -> id
 # ingredients_table: ingredient_id -> id, id -> recipe_id
 
-class RecData (Pluggable, BaseException):
+def db_url(file=None, custom_url=None):
```
Comment (maweki) [inline gourmet/backends/db.py]: please add type annotations while you're at it

File: gourmet/backends/db.py [hunk 2]
```diff
@@ -125,27 +133,27 @@ class RecData (Pluggable, BaseException):
     AMT_MODE_AVERAGE = 1
     AMT_MODE_HIGH = 2
 
-    _singleton = {}
+    _instance_by_db_url = {}
+
+    @classmethod
+    def instance_for(cls, file=None, custom_url=None):
```
Comment (maweki) [inline gourmet/backends/db.py]: please add type annotations to this method

File: gourmet/backends/db.py [hunk 3]
```diff
@@ -125,27 +133,27 @@ class RecData (Pluggable, BaseException):
...
+        url = db_url(file, custom_url)
+
+        if url not in cls._instance_by_db_url:
+            cls._instance_by_db_url[url] = cls(file, url)
 
-    def __init__ (self, file=os.path.join(gglobals.gourmetdir,'recipes.db'),
-                  custom_url=None):
+        return cls._instance_by_db_url[url]
+
+    def __init__ (self, file, url):
```
Comment (maweki) [inline gourmet/backends/db.py]: please add type annotations to this method

File: gourmet/backends/db.py [hunk 4]
```diff
@@ -1843,14 +1851,35 @@ def get_default_values (self, colname):
             return []
 
 
-class RecipeManager (RecData):
+class RecipeManager:
+    _instance_by_db_url = {}
 
-    def __init__ (self,*args,**kwargs):
+    @classmethod
+    def instance_for(cls, file=None, custom_url=None):
```
Comment (maweki) [inline gourmet/backends/db.py]: please add type annotations to this method

**Comment by takluyver** [comment]
I kind of feel that the `instance_for` annotations are a net negative, because the signature that previously fitted on one line is now an ugly mess over three lines. :shrug:

**Comment by maweki** [comment]
> I kind of feel that the `instance_for` annotations are a net negative, because the signature that previously fitted on one line is now an ugly mess over three lines. shrug

We will thank ourselves once we do any more complex refactoring. It looks good to me. Tank you.

**Comment by cydanil** [comment]
Thanks again :)

**Pull request merged**